consensus/quorum/one-node-staked-vote.go

```diff
@@ -101,38 +127,89 @@ func (v *stakedVoteWeight) Award(
 	return payout
 }
 
-func (v *stakedVoteWeight) UpdateVotingPower(staked shard.SlotList) {
-	s, _ := v.ShardIDProvider()()
+var (
+	errSumOfVotingPowerNotOne   = errors.New("sum of total votes do not sum to 100%")
+	errSumOfOursAndTheirsNotOne = errors.New(
+		"sum of hmy nodes and stakers do not sum to 100%",
+	)
+)
 
-	v.validatorStakes = map[shard.BlsPublicKey]stakedVoter{}
+func (v *stakedVoteWeight) SetVoters(
+	staked shard.SlotList,
+) (*TallyResult, error) {
+	s, _ := v.ShardIDProvider()()
+	v.voters = map[shard.BlsPublicKey]stakedVoter{}
 	v.Reset([]Phase{Prepare, Commit, ViewChange})
+	v.hmySlotCount = 0
+	v.stakedTotal = numeric.ZeroDec()
 
 	for i := range staked {
-		if staked[i].StakeWithDelegationApplied != nil {
-			v.validatorStakes[staked[i].BlsPublicKey] = stakedVoter{
-				true, false, staked[i].EcdsaAddress, *staked[i].StakeWithDelegationApplied,
-			}
-			v.total = v.total.Add(*staked[i].StakeWithDelegationApplied)
+		if staked[i].StakeWithDelegationApplied == nil {
+			v.hmySlotCount++
 		} else {
-			v.validatorStakes[staked[i].BlsPublicKey] = stakedVoter{
-				true, true, staked[i].EcdsaAddress, hSentinel,
-			}
+			v.stakedTotal = v.stakedTotal.Add(*staked[i].StakeWithDelegationApplied)
 		}
 	}
 
+	ourCount := numeric.NewDec(v.hmySlotCount)
+	ourPercentage := numeric.ZeroDec()
+	theirPercentage := numeric.ZeroDec()
+	totalStakedPercent := numeric.ZeroDec()
+
+	for i := range staked {
+		member := stakedVoter{
+			isActive:         true,
+			isHarmonyNode:    true,
+			earningAccount:   staked[i].EcdsaAddress,
+			effectivePercent: numeric.ZeroDec(),
+		}
+
+		// Real Staker
+		if staked[i].StakeWithDelegationApplied != nil {
+			member.isHarmonyNode = false
+			member.effectivePercent = staked[i].StakeWithDelegationApplied.
+				Quo(v.stakedTotal).
+				Mul(stakersShare)
+			theirPercentage = theirPercentage.Add(member.effectivePercent)
+		} else { // Our node
+			member.effectivePercent = numeric.NewDecFromBigInt(
+				v.QuorumThreshold(),
+			).Quo(ourCount)
+			ourPercentage = ourPercentage.Add(member.effectivePercent)
+		}
+
+		totalStakedPercent = totalStakedPercent.Add(member.effectivePercent)
+		v.voters[staked[i].BlsPublicKey] = member
+	}
+
 	utils.Logger().Info().
+		Str("our-percentage", ourPercentage.String()).
+		Str("their-percentage", theirPercentage.String()).
 		Uint32("on-shard", s).
-		Str("Staked", v.total.String()).
+		Str("Raw-Staked", v.stakedTotal.String()).
 		Msg("Total staked")
+
+	switch {
+	case totalStakedPercent.Equal(totalShare) == false:
```

Comment:
With the current code logic, these two cases are checking the same thing. totalStakedPercent = ours + theirs in the above logic. Due to precision problem, the final sum may not be 100%, we need to hard code it to make it 100%, such as giving the residual to the last voter to sum up to 100%

Reply:
the decimal wrapper was designed for this usage though, to take into account the precision problems.